app/bundles/LeadBundle/Form/Validator/Constraints/UniqueCustomFieldValidator.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mautic\LeadBundle\Form\Validator\Constraints;
+
+use Mautic\LeadBundle\Entity\Company;
+use Mautic\LeadBundle\Entity\Lead;
+use Mautic\LeadBundle\Model\CompanyModel;
+use Mautic\LeadBundle\Model\FieldModel;
+use Mautic\LeadBundle\Model\LeadModel;
+use Symfony\Component\Form\Form;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+class UniqueCustomFieldValidator extends ConstraintValidator
+{
+    private LeadModel $leadModel;
+    private CompanyModel $companyModel;
+    private FieldModel $fieldModel;
+
+    public function __construct(LeadModel $leadModel, CompanyModel $companyModel, FieldModel $fieldModel)
+    {
+        $this->leadModel    = $leadModel;
+        $this->companyModel = $companyModel;
+        $this->fieldModel   = $fieldModel;
+    }
+
+    /**
+     * @param Lead|Company|mixed $object
+     */
+    public function validate($object, Constraint $constraint): void
+    {
+        \assert($constraint instanceof UniqueCustomField);
+        \assert($object instanceof Lead || $object instanceof Company);
+
+        $form = $this->context->getRoot();
+        \assert($form instanceof Form);
+
+        $publishedUniqueFields = [];
+        $leadFields            = $this->fieldModel->getPublishedFieldArrays($constraint->object);
+        foreach ($leadFields as $field) {
+            if (false === $field['isPublished'] || $constraint->object !== $field['object']) {
+                continue;
+            }
+
+            if ($field['isUniqueIdentifer']) {
+                $publishedUniqueFields[] = $field['alias'];
+            }
+        }
+
+        $publishedUniqueFields = array_unique($publishedUniqueFields);
+
+        foreach ($publishedUniqueFields as $publishedUniqueField) {
+            if (!$form->has($publishedUniqueField)) {
+                continue;
+            }
+
+            $data = $form->get($publishedUniqueField)->getData();
+            if (null === $data || '' === $data) {
+                continue;
+            }
+
+            if ($object instanceof Lead && $this->isLeadFieldValid($object, $publishedUniqueField, $data)) {
```

Comment:
If there is more than one unique field then all unique fields should be passed together because there can be either OR or AND operator between the unique fields depending on the `company_unique_identifiers_operator` config option. This would work only for OR.

I was fighting the same problem in https://github.com/mautic/mautic/pull/11697

Reply:
Well, that's much more complicated than i thought. See last commit.

Reply:
Yep, I also spent more time than I expected on that detail. Thanks for attacking it!